### PR TITLE
FS-1406 friendly application id

### PR DESCRIPTION
**Pull request by harryyo**
### Added unique _reference_ attribute to application object

- Two new columns added to Application table/model: `key` and `reference`
- Multicolumn UniqueConstraint added to (`fund_id`, `round_id` and `key`)
- `key` is a random 6 (A-Z) character string which is defined on create_application() which when taken together with fund_id+round_id creates a unique application reference
- reference is a unique column on the applications table
- on create_application() a random `key` is selected and an attempt is made to create an application record with that key. If an existing record with an identical `fund_id`, `round_id` and `key` is found then a new random `key` is selected and a second attempt to create the application record is made using the new random `key`. This process is continued up to a `max_tries` of 10, until a an application record with a unique `reference` is successfully created or the `max_tries` is reached

### CODE REVIEW

File: db/migrations/versions/b2f22185655c_.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table("applications", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("key", sa.String(), nullable=False))
+        batch_op.add_column(
+            sa.Column("reference", sa.String(), nullable=False)
```
Comment (gidsg) [inline db/migrations/versions/b2f22185655c_.py:24]: I think we may need to make these Nullable else they will break for any existing applications in the DB which won't have those fields set. Have we tested this migration on the docker-runner?

File: api/routes/application/routes.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def post(self):
         round_id = args["round_id"]
         fund_id = args["fund_id"]
         empty_forms = ApplicationHelpers.get_blank_forms(fund_id, round_id)
-        application = ApplicationsMethods.create_application(
+        application = ApplicationsMethods().create_application(
```
Comment (SentientHousePlant) [inline api/routes/application/routes.py:39]: 🤔 Is this necessary? ApplicationsMethods only contains static methods?

File: tests/conftest.py [hunk 3]
```diff
@@ -35,6 +41,12 @@ def mock_get_data_fix(mocker):
     )
 
 
+@pytest.fixture()
```
Comment (SentientHousePlant) [inline tests/conftest.py:44]: 🤔 Does this need an autouse?

**Review by SentientHousePlant — CHANGES_REQUESTED** [review]

File: db/migrations/versions/b2f22185655c_.py (hunk 1 above)
Comment (harryyo) [reply db/migrations/versions/b2f22185655c_.py]: Yep hadn't got to the migrations yet. I think these should probably be not null fields after migration so I've added server_defaults during migration (for existing rows) and then removed the server_default straight after

**Review by donshellywork — APPROVED** [review]
lgtm

File: tests/test_routes.py [hunk 4]
```diff
@@ -44,6 +48,64 @@ def test_create_application_is_successful(client):
...
+        "round_id": "c603d114-5364-4474-a0c4-c41cbf4d3bbd",
+    }
+    response = client.post(
+        "/applications",
+        data=json.dumps(create_application_json),
+        content_type="application/json",
+        follow_redirects=True,
+    )
+    application = response.json
+    assert application["reference"][:8] == "COF-SUM-"
```
Comment (SentientHousePlant) [inline tests/test_routes.py:69]: ⛏️ You can use `startswith` and `endswith` for a slightly (very slightly) cleaner syntax.
https://docs.python.org/3/library/stdtypes.html#str.endswith

**Review by SentientHousePlant — APPROVED** [review]
LGTM! One small comment for you to look at.

**Review by SentientHousePlant — APPROVED** [review]
Lovely stuff Harry, lots of good refactoring and cleaning up too 💘

**Pull request merged**